base/range.jl

```diff
@@ -1480,3 +1480,111 @@ julia> mod(3, 0:2)  # mod(3, 3)
+a constant, calculated from the length.
+
+Like [`LinRange`](@ref), floating-point error may cause it to miss intermediate
+rational values, but the first and last elements should be exactly those provided.
+
+This is similar to `geomspace` in Python. Unlike `PowerRange` in Mathematica,
+you specify the number of elements not the ratio.
+Unlike `logspace` in Python and Matlab, the `start` and `stop` arguments are
+always the first and last elements of the result, not powers applied to some base.
+
+!!! compat "Julia 1.9"
+    This function requires at least Julia 1.9.
+
+# Examples
+```jldoctest
+julia> LogRange(10, 4000, 3)
+3-element LogRange{Float64}:
+ 10.0, 200.0, 4000.0
+
+julia> ans[2] ≈ sqrt(10 * 4000)  # middle element is the geometric mean
+true
+
+julia> LogRange(-62_500, -0.01, 5) |> collect  # endpoints are always exact
+5-element Vector{Float64}:
+ -62500.0
+  -1250.0
+    -25.0
+     -0.49999999999999994
+     -0.01
+
+julia> LogRange(√2, 32, 10)
+10-element LogRange{Float64}:
+ 1.41421, 2.0, 2.82843, 4.0, 5.65685, 8.0, 11.3137, 16.0, 22.6274, 32.0
+
+julia> LogRange(1, 1000, 4) ≈ 10 .^ (0:3)  # was logspace(0,3,4) in Julia < 1.0
+true
+
+julia> LogRange(1, -1+0im, 5) ≈ cis.(LinRange(0, pi, 5))  # complex numbers
+true
+```
+"""
+struct LogRange{T} <: AbstractArray{T,1}
+    start::T
+    stop::T
+    len::Int
+    function LogRange(start::T1, stop::T2, len::Integer) where {T1<:Number, T2<:Number}
+        T = float(promote_type(T1, T2))
+        lo = iszero(start) ? T(NaN) : T(start)
+        hi = iszero(stop) ? T(NaN) : T(stop)
+        if len < 0
+            throw(ArgumentError("LogRange can't have negative length"))
+        elseif len == 0
+            return new{T}(lo, hi, len)
+        elseif len == 1 && start != stop
+            throw(ArgumentError("endpoints of LogRange differ, with length = 1"))
+        end
+        if T <: Real && (start<0) ⊻ (stop<0)
+            throw(DomainError((start, stop),
+                "LogRange will only return complex results if called with a complex argument"))
+        end
+        new{T}(lo, hi, len)
+    end
+end
+
+size(r::LogRange) = (r.len,)
+
+first(r::LogRange) = r.start
+last(r::LogRange) = r.stop
+
+function getindex(r::LogRange, i::Integer)
+    @inline
+    @boundscheck checkbounds(r, i)
+    unsafe_getindex(r, i)
+end
+
+function unsafe_getindex(r::LogRange, i::Integer)
+    i isa Bool && throw(ArgumentError("invalid index: $i of type Bool"))
+    if r.len == 1
+        return r.start
+    end
+    if eltype(r) <: Real && r.start < 0  # constructor guarantees r.stop < 0 too
+        -loginterp(-r.start, -r.stop, Int(i), r.len)
+    else
+        loginterp(r.start, r.stop, Int(i), r.len)
+    end
+end
+
+function loginterp(lo::T, hi::T, j::Int, n::Int) where {T}
+    @inline
+    convert(T, lo^((n-j)/(n-1)) * hi^((j-1)/(n-1)))
+end
+
+function show(io::IO, r::LogRange)  # compact
+    print(io, "LogRange(")
```

Comment:
LinRange shows its first type parameter, even when compact.

Reply:
But not the second. My take here is that this is because it is ancient. Ideally either the numbers should print their types, or else the container:
```julia
julia> 0 * (1:2:3f0)  # hides container type, since 0.0f0 isa Float32
StepRangeLen(0.0f0, 0.0f0, 2)

julia> 1:2:3f0 |> tuple
(StepRangeLen(0.0f0, 0.0f0, 2),)

julia> ([1f0, 2f0],)  # here instead the container says Float32, numbers don't
(Float32[1.0, 2.0],)

julia> LinRange(1f0, 2f0, 3) |> tuple  # indicates Float32 twice, why?
(LinRange{Float32}(1.0f0, 2.0f0, 3),)
```
Perhaps `LinRange` should be modernised instead? Now that its length is allowed not to be an `Int` (as above), it is somewhat odd to rely on `3 isa Int` but not rely on `2.0f0 isa Float32`.

Note that this PR's display printing also differed from LinRange, by adding some space between the numbers -- again on the grounds that LinRange is ancient & we can do better. LinRange has since been modernised in #44937 to make exactly the same change, brining the two back into agreement.

Reply:
I agree that LinRange should not print redundant type info https://github.com/JuliaLang/julia/pull/47509

Reply:
BTW, at present there is no constructor `LogRange{Float64}(1,2,3)`, so printing that would certainly be a bad idea.

Whether there should be such a method, I don't know.